File: java/Dson-Codec/src/main/java/cn/wjybxx/dsoncodec/codecs/EnumCodec.java
/*
 * Copyright 2023-2024 wjybxx(845740757@qq.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package cn.wjybxx.dsoncodec.codecs;

import cn.wjybxx.base.EnumLite;
import cn.wjybxx.base.ObjectUtils;
import cn.wjybxx.dson.text.ObjectStyle;
import cn.wjybxx.dson.text.StringStyle;
import cn.wjybxx.dsoncodec.DsonCodecException;
import cn.wjybxx.dsoncodec.DsonObjectReader;
import cn.wjybxx.dsoncodec.DsonObjectWriter;
import cn.wjybxx.dsoncodec.TypeInfo;
import cn.wjybxx.dsoncodec.annotations.DsonCodecScanIgnore;
import cn.wjybxx.dsoncodec.annotations.DsonProperty;
import it.unimi.dsi.fastutil.ints.Int2ObjectMap;
import it.unimi.dsi.fastutil.ints.Int2ObjectOpenHashMap;

import javax.annotation.Nonnull;
import javax.annotation.Nullable;
import java.lang.reflect.Field;
import java.util.*;
import java.util.function.Supplier;

/**
 * 枚举编解码器
 * 如果枚举实现了{@link EnumLite}接口，则序列化时使用自定义数字，否则使用{@link Enum#ordinal()}
 *
 * @author wjybxx
 * date - 2023/4/24
 */
@DsonCodecScanIgnore
public final class EnumCodec<T extends Enum<T>> implements IEnumCodec<T> {

    private final Class<T> enumClass;
    private final EnumMap<T, EnumValueInfo<T>> value2EnumMap;
    private final Int2ObjectMap<EnumValueInfo<T>> number2EnumMap;
    private final Map<String, EnumValueInfo<T>> name2EnumMap; // 这允许为枚举指定别名

    /**
     * @param enumClass  枚举类
     * @param valueInfos 枚举值信息，允许自定义枚举序列化数据
     */
    public EnumCodec(Class<T> enumClass, List<EnumValueInfo<T>> valueInfos) {
        this.enumClass = Objects.requireNonNull(enumClass);

        this.value2EnumMap = new EnumMap<>(enumClass);
        this.number2EnumMap = new Int2ObjectOpenHashMap<>(valueInfos.size());
        this.name2EnumMap = HashMap.newHashMap(valueInfos.size());

        for (EnumValueInfo<T> valueInfo : valueInfos) {
            value2EnumMap.put(valueInfo.value, valueInfo);
            number2EnumMap.put(valueInfo.number, valueInfo);
            name2EnumMap.put(valueInfo.name, valueInfo);
        }
    }

    public EnumCodec(Class<T> enumClass) {
        this.enumClass = Objects.requireNonNull(enumClass);

        T[] enumConstants = enumClass.getEnumConstants();
        this.value2EnumMap = new EnumMap<>(enumClass);
        this.number2EnumMap = new Int2ObjectOpenHashMap<>(enumConstants.length);
        this.name2EnumMap = HashMap.newHashMap(enumConstants.length);

        Field[] enumFields = enumClass.getDeclaredFields();
        for (int idx = 0; idx < enumConstants.length; idx++) {
            T value = enumConstants[idx];
            int number = value instanceof EnumLite enumLite ? enumLite.getNumber() : value.ordinal();

            // 可通过注解指定DsonName -- java的枚举常量在所有其它字段之前
            DsonProperty annotation = enumFields[idx].getAnnotation(DsonProperty.class);
            String name;
            if (annotation != null && !ObjectUtils.isBlank(annotation.name())) {
                name = annotation.name();
            } else {
                name = value.name();
            }

            EnumValueInfo<T> valueInfo = new EnumValueInfo<>(value, number, name);
            value2EnumMap.put(valueInfo.value, valueInfo);
            number2EnumMap.put(valueInfo.number, valueInfo);
            name2EnumMap.put(valueInfo.name, valueInfo);
        }
    }

    @Override
    @Nullable
    public T forName(String name) {
        EnumValueInfo<T> valueInfo = name2EnumMap.get(name);
        return valueInfo == null ? null : valueInfo.value;
    }

    @Override
    @Nullable
    public T forNumber(int number) {
        EnumValueInfo<T> valueInfo = number2EnumMap.get(number);
        return valueInfo == null ? null : valueInfo.value;
    }

    @Override
    public String getName(T val) {
        EnumValueInfo<T> valueInfo = value2EnumMap.get(val);
        return valueInfo.name;
    }

    @Override
    public int getNumber(T val) {
        EnumValueInfo<T> valueInfo = value2EnumMap.get(val);
        return valueInfo.number;
    }

    @Nonnull
    @Override
    public TypeInfo getEncoderType() {
        return TypeInfo.of(enumClass);
    }

    /** false 可以将枚举简单写为整数 */
    @Override
    public boolean autoStartEnd() {
        return false;
    }

    @Override
    public void writeObject(DsonObjectWriter writer, T inst, TypeInfo declaredType, ObjectStyle style) {
        EnumValueInfo<T> valueInfo = value2EnumMap.get(inst);
        if (valueInfo == null) {
            throw new DsonCodecException("invalid enum value: %s, type: %s".formatted(inst, enumClass));
        }
        if (writer.options().writeEnumAsString) {
            writer.writeString(null, valueInfo.name, StringStyle.UNQUOTE);
        } else {
            writer.writeInt(null, valueInfo.number);
        }
    }

    @Override
    public T readObject(DsonObjectReader reader, Supplier<? extends T> factory) {
        if (reader.options().writeEnumAsString) {
            String name = reader.readString(null);
            EnumValueInfo<T> valueInfo = name2EnumMap.get(name);
            if (valueInfo != null) {
                return valueInfo.value;
            }
            throw new DsonCodecException("invalid enum value: %s, type: %s".formatted(name, enumClass));
        } else {
            int number = reader.readInt(null);
            EnumValueInfo<T> valueInfo = number2EnumMap.get(number);
            if (valueInfo != null) {
                return valueInfo.value;
            }
            throw new DsonCodecException("invalid enum value: %d, type: %s".formatted(number, enumClass));
        }
    }

}

File: csharp/Wjybxx.Dson.Codec/src/Codecs/EnumCodec.cs
#region LICENSE

// Copyright 2023-2024 wjybxx(845740757@qq.com)
// 
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
// 
//     http://www.apache.org/licenses/LICENSE-2.0
// 
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#endregion

using System;
using System.Collections.Generic;
using System.Reflection;
using Wjybxx.Commons;
using Wjybxx.Dson.Codec.Attributes;
using Wjybxx.Dson.IO;

namespace Wjybxx.Dson.Codec.Codecs
{
public readonly struct EnumValueInfo<T>
{
    public readonly T value;
    public readonly int number;
    public readonly string name;
    public readonly string numberString;

    public EnumValueInfo(T value, int number, string name) {
        this.value = value;
        this.number = number;
        this.name = name ?? throw new ArgumentNullException(nameof(name));
        this.numberString = EnumCodecUtil.ToString(number);
    }
}

/// <summary>
/// 默认枚举类的Codec
/// 注意：默认不支持序列化未在枚举中定义的枚举值 —— 其它特殊情况，建议直接使用int值。
/// </summary>
/// <typeparam name="T"></typeparam>
public sealed class EnumCodec<T> : IDsonCodec<T>, IKeyCodec<T> where T : struct, Enum
{
    private readonly Dictionary<T, EnumValueInfo<T>> _value2EnumDic;
    private readonly Dictionary<int, EnumValueInfo<T>> _number2EnumDic;
    private readonly Dictionary<string, EnumValueInfo<T>> _name2EnumDic; // 忽略大小写
    private readonly bool _isFlags;
    private readonly bool _isWriteAsString;

    /// <summary>
    /// 
    /// </summary>
    /// <param name="valueInfos">枚举值信息，允许自定义枚举序列化数据</param>
    /// <param name="isFlags">是否是Flags类型</param>
    public EnumCodec(List<EnumValueInfo<T>> valueInfos, bool? isFlags = null) {
        _value2EnumDic = new Dictionary<T, EnumValueInfo<T>>(valueInfos.Count);
        _number2EnumDic = new Dictionary<int, EnumValueInfo<T>>(valueInfos.Count);
        _name2EnumDic = new Dictionary<string, EnumValueInfo<T>>(valueInfos.Count, StringComparer.OrdinalIgnoreCase);
        _isFlags = isFlags ?? EnumCodecUtil.IsFlags<T>();
        _isWriteAsString = (EnumCodecUtil.GetEncodeFeatures<T>() & SerializeFeatures.EnumAsString) != 0;

        foreach (EnumValueInfo<T> valueInfo in valueInfos) {
            _value2EnumDic[valueInfo.value] = valueInfo;
            _number2EnumDic[valueInfo.number] = valueInfo;
            _name2EnumDic[valueInfo.name] = valueInfo;
        }
    }

    public EnumCodec() {
        T[] values = EnumUtil.GetValues<T>();
        string[] names = EnumUtil.GetNames<T>();
        _value2EnumDic = new Dictionary<T, EnumValueInfo<T>>(values.Length);
        _number2EnumDic = new Dictionary<int, EnumValueInfo<T>>(values.Length);
        _name2EnumDic = new Dictionary<string, EnumValueInfo<T>>(values.Length);
        _isFlags = EnumCodecUtil.IsFlags<T>();
        _isWriteAsString = (EnumCodecUtil.GetEncodeFeatures<T>() & SerializeFeatures.EnumAsString) != 0;

        FieldInfo[] enumFields = typeof(T).GetFields();
        for (int idx = 0; idx < values.Length; idx++) {
            T value = values[idx];
            int number = EnumUtil.GetIntValue(value);
            // 可通过注解指定DsonName -- 第一个元素是占位符，查询枚举关联的Field时需要+1
            DsonPropertyAttribute attribute = enumFields[idx + 1].GetCustomAttribute<DsonPropertyAttribute>();
            string name = attribute != null && !string.IsNullOrWhiteSpace(attribute.Name) ? attribute.Name : names[idx];

            EnumValueInfo<T> valueInfo = new EnumValueInfo<T>(value, number, name);
            _value2EnumDic[valueInfo.value] = valueInfo;
            _number2EnumDic[valueInfo.number] = valueInfo;
            _name2EnumDic[valueInfo.name] = valueInfo;
        }
    }

    #region 避免装箱

    public string EncodeKey(T value, SerializeFeatures features) {
        // 枚举Key必须存在对应的名字
        if (_value2EnumDic.TryGetValue(value, out EnumValueInfo<T> valueInfo)) {
            return (features & SerializeFeatures.EnumKeyAsString) != 0
                ? valueInfo.name
                : valueInfo.numberString;
        }
        throw new DsonCodecException($"invalid enum key: {value}, type: {typeof(T)}");
    }

    public T DecodeKey(string keyString) {
        // 枚举Key必须存在对应的名字 - Flags不应该作为Key
        if (int.TryParse(keyString, out int number)) {
            if (number == 0) {
                return default;
            }
            if (_number2EnumDic.TryGetValue(number, out EnumValueInfo<T> valueInfo)) {
                return valueInfo.value;
            }
        } else {
            if (_name2EnumDic.TryGetValue(keyString, out EnumValueInfo<T> valueInfo)) {
                return valueInfo.value;
            }
        }
        throw new DsonCodecException($"invalid enum key: {keyString}, type: {typeof(T)}");
    }

    #endregion

    public void WriteObject(IDsonObjectWriter writer, T inst, Type declaredType, SerializeFeatures features) {
        if (!_value2EnumDic.TryGetValue(inst, out EnumValueInfo<T> valueInfo)) {
            writer.WriteInt(EnumUtil.GetIntValue(inst)); // 可能是default或flags
            return;
        }
        bool isWriteAsString = _isWriteAsString || IsWriteAsString(features, writer);
        if (isWriteAsString) {
            writer.WriteString(valueInfo.name, SerializeFeatures.StringUnquote);
        } else {
            writer.WriteInt(valueInfo.number);
        }
    }

    public T ReadObject(IDsonObjectReader reader, Type declaredType, DeserializeFeatures features, Func<object>? factory = null) {
        if (reader.CurrentDsonType.IsNumber()) {
            int number = reader.ReadInt();
            if (number == 0) {
                return default;
            }
            if (_number2EnumDic.TryGetValue(number, out EnumValueInfo<T> valueInfo)) {
                return valueInfo.value;
            }
            if (_isFlags) {
                return (T)Enum.ToObject(typeof(T), number);
            }
            throw new DsonCodecException($"invalid enum value: {number}, type: {typeof(T)}");
        }
        if (reader.CurrentDsonType == DsonType.String) {
            string name = reader.ReadString();
            if (_name2EnumDic.TryGetValue(name, out EnumValueInfo<T> valueInfo)) {
                return valueInfo.value;
            }
            if (name.Contains('|')) { // Flags格式
                int number = ParseFlags(name);
                return (T)Enum.ToObject(typeof(T), number);
            }
            throw new DsonCodecException($"invalid enum value: {name}, type: {typeof(T)}");
        }
        if (reader.CurrentDsonType == DsonType.Array) {
            int number = ReadArray(reader);
            if (number == 0) {
                return default;
            }
            return (T)Enum.ToObject(typeof(T), number);
        }
        throw DsonIOException.InvalidDsonType(new List<DsonType>()
        {
            DsonType.Int32, DsonType.String, DsonType.Array
        }, reader.CurrentDsonType);
    }

    private int ReadArray(IDsonObjectReader reader) {
        int value = 0;
        while (reader.ReadDsonType() != DsonType.EndOfObject) {
            string name = reader.ReadString();
            if (_name2EnumDic.TryGetValue(name, out EnumValueInfo<T> valueInfo)) {
                value |= valueInfo.number;
                continue;
            }
            throw new DsonCodecException($"invalid enum value: {name}, type: {typeof(T)}");
        }
        return value;
    }

    private int ParseFlags(string str) {
        int value = 0;
        foreach (string name in ObjectUtil.SplitAndTrim(str, '|')) {
            if (_name2EnumDic.TryGetValue(name, out EnumValueInfo<T> valueInfo)) {
                value |= valueInfo.number;
                continue;
            }
            throw new DsonCodecException($"invalid enum value: {name}, type: {typeof(T)}");
        }
        return value;
    }

    private bool IsIgnoreCase(DeserializeFeatures features, IDsonObjectReader reader) {
        if ((features & DeserializeFeatures.EnumIgnoreCase) != 0) return true;
        TypeMeta typeMeta = reader.ContainerTypeMeta;
        if (typeMeta != null) {
            features = typeMeta.decodeFeatures;
            if ((features & DeserializeFeatures.EnumIgnoreCase) != 0) return true;
        }
        features = reader.Options.decodeFeatures;
        return (features & DeserializeFeatures.EnumIgnoreCase) != 0;
    }

    private bool IsWriteAsString(SerializeFeatures features, IDsonObjectWriter writer) {
        if ((features & SerializeFeatures.EnumAsString) != 0) return true;
        if ((features & SerializeFeatures.EnumAsNumber) != 0) return false;
        TypeMeta typeMeta = writer.ContainerTypeMeta;
        if (typeMeta != null) {
            features = typeMeta.encodeFeatures;
            if ((features & SerializeFeatures.EnumAsString) != 0) return true;
            if ((features & SerializeFeatures.EnumAsNumber) != 0) return false;
        }
        features = writer.Options.encodeFeatures;
        return (features & SerializeFeatures.EnumAsString) != 0;
    }
}
}